### ISSUE TITLE
Implement TypeScript in New Expensify

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#16665]
### 📄 Link to the design doc
[docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#)

### 🖇️ Other links

- [TypeScript configurations](https://docs.google.com/document/d/1YczjOMTEOdbWhsn0nEC2MzA31ME7OsNw-DYhIEJ4Hk8/edit#)
- [File by file migration plan](https://docs.google.com/document/d/1KGimBQh0cLg7yBTQPODLMtEMR2yCPa8IEdOkUQS7NdU/edit#)

### Strategy
Our strategic roadmap for the next couple years is Reunification – moving all of our existing users from OldDot to NewDot, and it’s going to look like a completely new product. During this transition period, there’s an increased risk for churn, and stability of the platform is more important than ever. Any lines of defense between our customers and bad code will pay dividends during any tumultuous times to come.

### Problem
JavaScript has been very popular for front-end development for a long time, and over the years it has evolved with new developer-friendly features that make it easier to GSD. Overall, it’s pretty great. However, it is still plagued by one fundamental problem: runtime errors stemming from incorrect types. This common class of problem leads to bugs, ranging in severity from showing weird things on the screen to crashing the whole app.

While there are many bugs that can be caused by incorrect types, the most common example of a runtime type safety error is `Cannot access property X of undefined`. A [quick GitHub search](https://github.com/search?q=repo%3AExpensify%2FApp+%22of+undefined%22&type=issues) reveals that over the last couple years we have introduced hundreds of this type of bug into the NewDot codebase, each of them stemming from this same root cause – attempting to access a property on a variable whose value is `undefined`. When this occurs, the app crashes.

### Solution
Let's commit to adopting TypeScript in New Expensify, an awesome superset of JavaScript that gives us the benefits of using a compiled language like C++! That way, type errors can be caught at compile-time instead of run-time, and a whole class of common bugs will be eliminated from our product. #BugZero here we come.

Furthermore, let’s engage the near-limitless capacity of our open-source community to do it in parallel to our existing roadmap, without pausing any other feature development in the meantime.

## Tasks

- [x] Post **Proposal (full Problem/Solution statement)** in `#expensify-open-source`
- [x] Wait at least one full business day, and until the post has a majority (2/3) of positive reactions (👍)
- [x] Paste Proposal in the space above with a link to the Slack thread
- [x] Email `strategy@expensify.com` and paste in the Proposal
- [x] Fill out the **High-level overview of the problem**, **Timeline**, and **Terminology** sections of the Design Doc
- [x] Email `strategy@expensify.com` (continue the same email chain as before) with the link to your Design Doc
- [x] Host a **pre-design** meeting ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1665794669539419)) in `#expensify-open-source` to discuss any necessary details in public before filling out the **High-level of proposed solution** section.
- [x] Fill out the **High-level of proposed solution** section
- [x] Email `stategy@expensify.com` again with links to the doc and pre-design conversation in Slack
- [x] Add the `DesignDocReview` label to get the **High-level of proposed solution** section reviewed
- [x] Respond to any questions or concerns and bring up blockers in Slack to get a consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Host another pre-design meeting in `#expensify-open-source` to ask for engineering feedback on the technical solution.
- [x] Fill out the **Detailed implementation of the solution** and related sections.
- [x] Re-add the `DesignDocReview` label to this issue
- [x] Respond to any questions or concerns and bring up blockers in Slack to get consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Email `strategy@expensify.com` one last time to let them know the Design Doc is moving into the implementation phase
- [x] Implement the changes
- [ ] Send out a follow up email to `strategy@expensify.com` once everything has been implemented and do a **Project Wrap-Up** retrospective that provides:
    - Summary of what we accomplished with this project
    - What went well?
    - What could we have done better?
    - What did we learn?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`NewFeature`), see https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=BugZero%20process%20steps%20for%20feature%20requests for more details.

**Comment by roryabraham** [comment]
Here's a high-level overview of what I'm imagining for the rollout plan:

```mermaid
flowchart LR
    A(Setup TypeScript for E/App source code)
    X(Setup TypeScript for GitHub Actions)
    Y(Setup TypeScript for Jest)

    B(Laying Foundation)
    C(Materials and Plumbing)
    D(Construction)
    E(Finishings)
    
    F(Implement TypeScript in GitHub Actions)
    G(Implement TypeScript for Jest tests)
    
    subgraph Main App
        A --> B --> C --> D --> E
    end
    
    subgraph GitHub Actions
        direction LR
        X --> F
    end
    
    subgraph Jest
        Y ----> G
        C --> G
    end
    
    H(Setup TypeScript in react-native-onyx)
    I(Implement TypeScript in react-native-onyx)
    
    subgraph react-native-onyx
        direction LR
        H --> I
    end
```

Most of this is pretty self-explanatory, but I'll go into some extra detail for the rollout plan for the main source code.

- **Setup:**
    - Add TypeScript tooling for local development and in the deploy process.
    - Migrate App.js to App.ts
- **Laying Foundation:**
    - Create type definitions for key entities (such as `Report` and `Policy`)
    - Migrate simple, widely used files that will use mostly inferred types (such as `CONST` and `ONYXKEYS`)
    - After this phase, _disallow new JS files_
- **Materials and Plumbing:**
    - Migrate low-level components and utilities that don’t depend much on other files. Because these files generally don’t have many dependencies, it’s less likely that we’ll scope creep, have tons of conflicts, or find that they need to be on HOLD for other migrations.
    - Files that don’t have other dependencies are like “materials”. Starting with most of the basic materials in place will make the rest of the construction go faster and smoother.
    - However, we won’t delay the main “construction phase” to get _all_ the materials in place, just the most common and critical ones.
- **Construction:**
   - Migrate our most commonly used and most critical components and libraries.
- **Finishings:**
   - Finish migrating, until there are no JS files in our App source code
   
We can use [`deprank`](https://github.com/codemix/deprank) to determine the number of dependencies and dependents on a file-by-file basis.

**Comment by roryabraham** [comment]
Posted a second predesign [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1681835514949819) yesterday

**Comment by roryabraham** [comment]
Started working on a doc here: https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#

**Comment by neil-marcellini** [comment]
I'm going to help Rory with this by asking external agencies to collaborate on a pre-design / writing the doc, and then making sure the process is going well.

**Comment by neil-marcellini** [comment]
I didn't get to this today.

**Comment by neil-marcellini** [comment]
I posted to the Callstack and Software Mansion Slack channels asking for proposals on how they would plan and implement this project.

https://expensify.slack.com/archives/C03UK30EA1Z/p1684525440666989
https://expensify.slack.com/archives/C04878MDF34/p1684525456740769

**Comment by neil-marcellini** [comment]
I got a reply from Callstack and SWM wants a day or two to post something.

**Comment by neil-marcellini** [comment]
I created [#typescript-new-expensify](https://expensify.slack.com/archives/C0593T50UHG) and invited CK and SWM. @hayata-suenaga is helping out now too!

I still need to read through the proposal that software mansion already wrote, and then I’ll let them know that we’re waiting for their joint proposal and ask on an ETA for that. Maybe Hayata could tackle that?

**Comment by hayata-suenaga** [comment]
@neil-marcellini thank you for adding me to this issue

I thought you asked for proposals separately from Software Mansion and CallStack. Do they know they have to produce a joint proposal?

**Comment by hayata-suenaga** [comment]
okay CallStack is willing to collaborate

>2. Proposal: We will review [the proposal](https://expensify.slack.com/archives/C01GTK53T8Q/p1679920586360439) published in March by Fabio and other proposals and comments from contributors to gather the most relevant information and update the proposal and plan in cooperation with SWM

[from Tomek](https://expensify.slack.com/archives/C03UK30EA1Z/p1684770041850859?thread_ts=1684525440.666989&cid=C03UK30EA1Z)

**Comment by hayata-suenaga** [comment]
and Software Mansion is saying they're already in touch with CallStack

> This is our proposal on how to tackle migration, and we are already in touch with Callstack to compile a common one

[From Michal](https://expensify.slack.com/archives/C04878MDF34/p1684856644801569?thread_ts=1684525456.740769&cid=C04878MDF34)

**Comment by hayata-suenaga** [comment]
@neil-marcellini do you know if we have access to the collaboration channel between CallStack and Software Mansion?

I have no idea how they're communicating with each other, but I think we should have access to their communication channel.

Also, just curious. Why have we decided to use two agencies for the this migration project instead of just one?

**Comment by hayata-suenaga** [comment]
I just realized we forgot to consider the migration of [`expensify-commons`](https://github.com/Expensify/expensify-common) in [this diagram](https://github.com/Expensify/App/issues/16665#issuecomment-1491735911)

react-native-onyx has a dependency for `expensify-commons` so I think we have to migrate this repo first

**Comment by roryabraham** [comment]
I don't think there were any plans to migrate expensify-common, because:

- It's still used by OldDot, so we'd need to make it consumable by non-ts dependencies (which is possible but presents extra complications
- I'm not sure we'll even keep expensify-common forever, because we're working to deprecate all of our old front-end code, and eventually all our JS would live in E/App and we wouldn't need expensify-common anymore.

Anyways, TypeScript can import and use JS-only libs, so we definitely don't have to migrate expensify-common to TypeScript before react-native-onyx or E/App

**Comment by hayata-suenaga** [comment]
ah we're just using two Str methods inside `react-native-onyx` we can even remove them and `expensify-common` dependency during the migration from onyx

**Comment by hayata-suenaga** [comment]
we're also using only `Str` util class and `ExpensiMark` in App. we can definitely remove `expensify-common` them down the line from App, too.

there might be extra setup necessary to use a library that doesn't have a type declarations though. I guess vendors will gonna figure out that one

**Comment by mountiny** [comment]
> Also, just curious. Why have we decided to use two agencies for the this migration project instead of just one?

Because we are closely working with both of them and they both have lots of experience and this project has quite a bit of config/ setup decisions made so its ideal if they can both (ideally also Margelo will review the plan) with agree on the next steps. Everyone will be happy with the set up in future ad they can have their say now.

**Comment by hayata-suenaga** [comment]
@aleqsio presented [a list of suggestions](https://expensify.slack.com/archives/C0593T50UHG/p1685179390610999) on slack on how we type data in Onyx. I'll look at them and ask further questions to them.

**Comment by neil-marcellini** [comment]
I posted [an update](https://expensify.slack.com/archives/C0593T50UHG/p1685493243952829) in the channel to make it clear that the next step is to finalize the high level of the doc and get it reviewed.

**Comment by hayata-suenaga** [comment]
I also provided [the time estimation](https://expensify.slack.com/archives/C0593T50UHG/p1685515697839649) for the reviews to take. It usually takes around one week for a review to be completed right?

vendors probably want to know when they can start working so that they can allocate their team members accordingly

**Comment by hayata-suenaga** [comment]
waiting for Błażej (tech lead from SWM) to review the high level portion of the design doc. Once they think it's fine, we can them it for the internal view.

**Comment by maddylewis** [comment]
I have read and reviewed this Design Doc!

**Comment by mountiny** [comment]
I have read and reviewed this Design Doc!

**Comment by NickTooker** [comment]
I have read and reviewed this Design Doc!

**Comment by iwiznia** [comment]
I have read and reviewed this Design Doc!

**Comment by puneetlath** [comment]
I have read and reviewed this Design Doc!

**Comment by Gonals** [comment]
I have read and reviewed this Design Doc!

**Comment by Julesssss** [comment]
I have read and reviewed this Design Doc!

**Comment by flaviadefaria** [comment]
I have read and reviewed this Design Doc!

**Comment by JmillsExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by hayata-suenaga** [comment]
All the required reviews are done for the high level 🎊 

Now the next step is:

- Finalize the TS guideline section of the design doc
- Send the detailed section review

**Comment by hayata-suenaga** [comment]
## Links
- [TS config discussion](https://docs.google.com/document/d/1YczjOMTEOdbWhsn0nEC2MzA31ME7OsNw-DYhIEJ4Hk8/edit#heading=h.qqrdq15cod3f)
- [Feature grouping discussion](https://docs.google.com/document/d/1KGimBQh0cLg7yBTQPODLMtEMR2yCPa8IEdOkUQS7NdU/edit#heading=h.i8ucryvx64pq)

**Comment by PauloGasparSv** [comment]
I have read and reviewed this Design Doc!

**Comment by flodnv** [comment]
I have read and reviewed this Design Doc!

**Comment by AndrewGable** [comment]
I have read and reviewed this Design Doc!

**Comment by twisterdotcom** [comment]
I have read and reviewed this Design Doc!

**Comment by neil-marcellini** [comment]
I "reviewed" the detailed design (as an author) and left a few comments. I think Hayata will address those and then send the detailed section out as soon as the high level review is complete.

**Comment by hayata-suenaga** [comment]
### 🔗🔗 🔗   Detailed section link

https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#

**Comment by flaviadefaria** [comment]
I have read and reviewed this Design Doc!

**Comment by Li357** [comment]
I have read and reviewed this Design Doc!

**Comment by danielrvidal** [comment]
Is this a super high priority? It looks like it was created with a Daily tag for everyone to review in one day, which normally we reserve for things that are impacting the next couple waves. So I just wanted to make sure this applies to this too as the urgency in the doc talks about no timeline.

**Comment by francoisl** [comment]
I have read and reviewed this Design Doc!

**Comment by AndrewGable** [comment]
I have read and reviewed this Design Doc!

**Comment by hayata-suenaga** [comment]
> Is this a super high priority? It looks like it was created with a Daily tag for everyone to review in one day, which normally we reserve for things that are impacting the next couple waves. @danielrvidal

This issue has the "daily" label because this effort involves external vendors who already started working on some tasks. Migration is one of those projects that don't have a clear deadline. But they're also the kind of projects that benefit from a short lifespan.

As this design dos is rather short, we thought we could justify requesting reviews with the daily label 🙇

**Comment by flodnv** [comment]
I have read and reviewed this Design Doc!

**Comment by puneetlath** [comment]
I have read and reviewed this Design Doc!

**Comment by maddylewis** [comment]
I have read and reviewed this Design Doc!

**Comment by NickTooker** [comment]
I have read and reviewed this Design Doc!

**Comment by danielrvidal** [comment]
I have read and reviewed this Design Doc!

**Comment by PauloGasparSv** [comment]
I have read and reviewed this Design Doc!

**Comment by twisterdotcom** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
We've seen good activity and detailed reviews have been moving along nicely.

**Comment by hayata-suenaga** [comment]
The detailed review is done 🎉

Here are the next steps. For the assignment and division of tasks and tracking of subtasks of these four tasks, we delegate them to SWM and CallStack. Please communicate with each other to divide responsibilities 🙇 @fabioh8010 @blazejkustra

For each of these four tasks, I created a GH issue. We'll eventually put these issues inside a GH Project we'll create for the TypeScript migration. Please report any progress in the respective issue.

I provided rough deadlines for each task. These are more like guidelines rather than strict deadlines. Please provide your own estimate on the completion time of the task in the issue itself.

### 1. [Define TypeScript guidelines](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.fdd1mo9igwgk)

    Who: SWM and CallStack.
    Review: Once done, post this to the `engineering-chat` internal Slack channel to have the internal team review and learn TypeScrit usages.
    Rough deadline: We want to send the guideline for review by internal engineers by the end of this week.
    
    A rough overview of the guideline is in [the design doc](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.pz75s9ij30u2). Let's create the first version of the guideline in `app/contributingGuides/TYPESCRIPT_STYLE.md`. Let's add succinct code examples as needed. Create sub-files if necessary to divide the guideline into sections.
  
    Please write the guideline following conversions widely used and also those used already in other guidelines in the repo (ex. usage of shall, should, or must).

    GitHub: https://github.com/Expensify/App/issues/20623

### 2. [Prepare libraries for migration](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.59gvew7hbfe5)

  Who: SWM and CallStack.
  Rought deadline: By mid-next week if possible (I don't know how complicated it will be to creat type definitions for onyx)

  Create type definitions for react-native-onyx and expensify-commons. Until these definitions are complete, we cannot move to the migration of App.
  
  - react-native-onyx
    GH isssue: https://github.com/Expensify/expensify-common/issues/547
    
  - expensify-common
    GH issue: https://github.com/Expensify/expensify-common/issues/547

### 3. [Prepare the GH project and issues](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.y4uk5ut31hte)

  Who: SWM and CallStack.
  @fabioh8010 is already testing GitHub projects and figuring out how to create migration issues.

  This task doesn't have a corresponding GH issue.

### 4. [Setup App for TypeScript](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.afxe5t69grn4)

  Who: SWM and CallStack.
  [PoC PR](https://github.com/Expensify/App/pull/20179/files#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5L6) is getting quite complete. Let's try to finish this by this weekend if possible.
  GH issue: https://github.com/Expensify/App/issues/20626

**Comment by neil-marcellini** [comment]
The PR to set up the app for typescript is under review and I just finished my review. I'm eager to keep helping out mostly by reviewing.

**Comment by roryabraham** [comment]
The main PR to integrate TypeScript toolchain into E/App is in review. Exciting!

**Comment by neil-marcellini** [comment]
I started reviewing the [Typescript guidelines](https://github.com/Expensify/App/pull/21050) PR today.

**Comment by blazejkustra** [comment]
PR with type definitions for `expensify-common` is ready [here](https://github.com/Expensify/expensify-common/pull/553). Please check it out! 👀

**Comment by hayata-suenaga** [comment]
### Latest update

- RP for adding type definition files for `expensify-common` is under review
- GH project for TS migration was created. (it's still private, but the visibility will be changed to public soon)
- @mountiny authored a PR for a script that creates GH issues for all migration tasks
- TS guideline is under review. @hayata-suenaga is going to review it this weekend.
- PR for setting up configurations for TypeScript on NewDot is under review and almost ready

### Next Steps

- Start working on writing type definition files for `react-native-onyx`
- CallStack and SWM dividing work for the initial set of migration tasks (migration of leaf files)
- Publish TypeScript guideline
- Post a slack comment about TypeScript migration on #expensify-open-source channel
    - Tell contributors not to post proposals to migration issues yet

**Comment by hayata-suenaga** [comment]
The TypeScript GH Project now has the public visibility

**Comment by hayata-suenaga** [comment]
[A new PR](https://github.com/Expensify/App/pull/21549) for TS guidelines was created, and I expect the review to be finished by the end of this week.

@fabioh8010 is working on creating type definition files for `react-native-onyx`

script for creating GH migration issues is ready

We should start thinking about typing core data. The PoC PR for this is already made [here](https://github.com/Expensify/App/pull/19966) by @thiagobrez.

**Comment by neil-marcellini** [comment]
We're planning to merge the [expensify-common PR](https://github.com/Expensify/expensify-common/pull/553) by EOD and the [App set up PR](https://github.com/Expensify/App/pull/20179) by EOW.

**Comment by hayata-suenaga** [comment]
- Finalizing the decision regarding the module organization strategy [here](https://expensify.slack.com/archives/C0593T50UHG/p1687985968639359?thread_ts=1687768826.156249&cid=C0593T50UHG)

- Asked the progress on typing core pieces of data on [slack](https://expensify.slack.com/archives/C0593T50UHG/p1687986763818579)

**Comment by cead22** [comment]
I have read and reviewed this Design Doc!

**Comment by hayata-suenaga** [comment]
### Current progress

- [`react-native-onyx` PR](https://github.com/Expensify/react-native-onyx/pull/267) is in progress.
- [`expensify-common PR`](https://github.com/Expensify/expensify-common/pull/553) has been merged.
- [TS setup PR](https://github.com/Expensify/App/pull/20179) is finalizing with additional ESLint rules
- [TS guideline PR](https://github.com/Expensify/App/pull/21549) is almost ready
- [Typing core data](https://expensify.slack.com/archives/C0593T50UHG/p1688138005327889?thread_ts=1687986763.818579&cid=C0593T50UHG) is WIP
- [script for creating migration GH issues](https://github.com/Expensify/Web-Expensify/pull/37920) are ready

**Comment by hayata-suenaga** [comment]
Requested internal engineers' review on [the core data PR](https://expensify.slack.com/archives/C0593T50UHG/p1688569554410559?thread_ts=1687986763.818579&cid=C0593T50UHG) in NewDot [here](https://expensify.slack.com/archives/C03TQ48KC/p1688582372669129).

**Comment by roryabraham** [comment]
Progress is ongoing. We were discussing a related change (that can happen separately from the main TS migration) in these threads:

- https://expensify.slack.com/archives/C0593T50UHG/p1688115737721429
- https://expensify.slack.com/archives/C01GTK53T8Q/p1688767129082749

I personally think the most significant decisions that are being made are how we are typing react-native-onyx, and I sadly am not up-to-date on the latest there

**Comment by hayata-suenaga** [comment]
> the most significant decisions that are being made are how we are typing react-native-onyx, and I sadly am not up-to-date on the latest there

@roryabraham 

The draft PR for creating definition files in Onyx is going [here](https://github.com/Expensify/react-native-onyx/pull/267).

**Comment by hayata-suenaga** [comment]
will update tonight

**Comment by hayata-suenaga** [comment]
# End of week update

- [The guideline PR](https://github.com/Expensify/App/pull/21549) is done and waiting everyone's approval.
- [PR](https://github.com/Expensify/react-native-onyx/pull/267) for adding type definition for Onyx is progressing nicely.
  We don't know if we want to migrate to `useOnyx` along with the TS migration. The discussion is happening [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1689356214581179)
- [PR for tying Onyx data is being reviewed](https://github.com/blazejkustra/App/pull/1)

### The following PRs are waiting for the migration to start
- [TS setup for App](https://github.com/Expensify/App/pull/20179)
- [Script for creating migration GH issues](https://github.com/Expensify/Web-Expensify/pull/37920)
- [De-structure props](https://github.com/Expensify/App/issues/22671)

### Action items for next week
- Decide if we migrate to `useOnyx` during TS migration
- Finish TS-Onyx PR

**Comment by hayata-suenaga** [comment]
# End of week update

- [The guideline PR](https://github.com/Expensify/App/pull/21549) is done and waiting everyone's approval.
- [PR](https://github.com/Expensify/react-native-onyx/pull/267) for adding type definition for Onyx is progressing nicely.
  We don't know if we want to migrate to `useOnyx` along with the TS migration. The discussion is happening [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1689356214581179)
- [PR for tying Onyx data is being reviewed](https://github.com/blazejkustra/App/pull/1)

### The following PRs are waiting for the migration to start
- [TS setup for App](https://github.com/Expensify/App/pull/20179)
- [Script for creating migration GH issues](https://github.com/Expensify/Web-Expensify/pull/37920)
- [De-structure props](https://github.com/Expensify/App/issues/22671)

### Action items for next week
- Decide if we migrate to `useOnyx` during TS migration
- Finish TS-Onyx PR

**Comment by roryabraham** [comment]
Had a sync with @hayata-suenaga today. The main action item at the moment is getting all the reviews in for [this big PR](https://github.com/blazejkustra/App/pull/1)

**Comment by hayata-suenaga** [comment]
# Mid week update

- TS guideline is almost ready. Lets merge it by the end of this week and announce it on the open source channel so that contributors will have time to look over the guideline before the migration starts (migration tasks will be only opened to contributors in the later stage of the migration, but they might encounter merge conflicts and have to edit TypeScript code)

- [Onyx + TS PR](https://github.com/Expensify/react-native-onyx/pull/267) is still being actively reviewed and worked on. Let's aim to merge this by the end of Wednesday next week 👍 

- @blazejkustra let us know if you need more help from internal engineers for [the core Onyx data + TS PR](https://github.com/blazejkustra/App/pull/1). Seems like all files are reviewed by internal engineers at least once.

- Let's run [the script](https://github.com/Expensify/Web-Expensify/pull/37920) to create all migration tasks once the Onyx + TS PR and core Onyx data + TS PR are merged next week.

**Comment by mountiny** [comment]
> @roryabraham , @mountiny, and @neil-marcellini do you have any idea how we can make sure that engineers are reminded to update the TS type definitions for Onyx data when the backend data shapes change?

I am not sure, maybe some action/ webhook analyzing the code changes

@blazejkustra @fabioh8010 any recommendations?

**Comment by blazejkustra** [comment]
> > @roryabraham , @mountiny, and @neil-marcellini do you have any idea how we can make sure that engineers are reminded to update the TS type definitions for Onyx data when the backend data shapes change?
> 
> I am not sure, maybe some action/ webhook analyzing the code changes
> 
> @blazejkustra @fabioh8010 any recommendations?

So, there are couple of ways to do it:

1. **Custom-Built Codegen**: Although we don't use GraphQL on the backend, we can draw some inspiration from GraphQL. The concept is to implement a code generator, which could automatically generate TS type definitions based on our backend data shape. Writing a custom codegen for a PHP backend might be a challenging task. Such approach might require some investment of time and resources initially but has the potential to make the backend-frontend sync easier in the longer run.
2. **Publicly Defined Schema**: Similar approach, publicly define the schema and utilize it at both ends - backend and frontend. [JSON Schema](https://json-schema.org/) could be utilized for that. To ensure that backend is in sync with the schema, input and output of every request would be validated with the schema. On the frontend we could either create types manually and validate incoming data in the runtime or use [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript) to generate TS types automatically.
3. **New Development Process**: New step in the review process, whenever backend data shapes are changed require a change on the frontend. That could be backend developer responsibility to make sure that changes are reflected in the types on frontend. Couple ideas:
    - Additional checkbox in checklist on the backend PR
    - Automation that checks data shapes on each commit on the backend, if needed require a frontend change (not sure if that's possible because I don't have access to the backend)
    - Runtime validation in the App, so that types defined in the App are compared with what we get from the API ([Zod](https://github.com/colinhacks/zod) could be useful for that)

**Comment by hayata-suenaga** [comment]
@mountiny I was asking how we can change the internal process (add new checklist item, etc) for Web-E PRs so that engineers are reminded to change the type definition of the backend data changes.

should we just edit the PR template for Web-E? Should we also add a checklist for `Auth`, too?

**Comment by mountiny** [comment]
@hayata-suenaga I am curious though if checklist will be enough to check for this. I think we can start with it and see if we will have people ignoring it

**Comment by hayata-suenaga** [comment]
@mountiny yes I think [the option #2](https://github.com/Expensify/App/issues/16665#issuecomment-1645551437) @blazejkustra suggested is something we want to consider down the line

but I'm thinking something we can implement right now that at least do something toward ensuring type consistency between front end and backend.

I gonna create issues for updating PR templates and if I have time, I gonna make PRs for them today

**Comment by roryabraham** [comment]
> @roryabraham , @mountiny, and @neil-marcellini do you have any idea how we can make sure that engineers are reminded to update the TS type definitions for Onyx data when the backend data shapes change?

I think we should punt this problem (or maybe rumor of a problem) for now and just focus on getting TypeScript built out in a leaner V1. Long-term I love the idea of having a shared, versioned schema for Onyx used by both the front-end and back-end to automate the compatibility between layers. But I feel strongly that we're getting ahead of ourselves here.

We don't currently have _anything_ to ensure Onyx schemas are compatible in the front-end and back-end, and it's been working mostly ok. The amount of type safety we'll get in the front-end alone will be a huge win.

FWIW, if someone's updating something in Onyx in PHP, they'll probably try to access that property in E/App, and notice that the type is wrong. I kind of think that's enough to start.

**Comment by roryabraham** [comment]
That said, if we're going to do anything I agree with @mountiny's suggestion that we can add a webhook comment like:

> I noticed that you changed some code that has to do with Onyx. Did you adjust the Onyx schema? If so, make sure you adjust the corresponding Onyx type declaration in E/App.

We could probably do that pretty easily but I don't think it should be a top priority for this project.

**Comment by hayata-suenaga** [comment]
Will merge the guideline PR this afternoon and post the announcement

**Comment by neil-marcellini** [comment]
I'm continuing to review the PRs to type Onyx data in App and to add type declarations for Onyx. Both are very close I think.

**Comment by hayata-suenaga** [comment]
### Mid-week update

- [The TS guideline PR](https://github.com/Expensify/App/pull/21549) has been merged, and announcements are posted in [the `#expensify-open-source`](https://expensify.slack.com/archives/C01GTK53T8Q/p1690233841812369) and [`#engineering-chat` (internal)](https://expensify.slack.com/archives/C03TQ48KC/p1690306494923019) slack channels
- [App TS Setup PR](https://github.com/Expensify/App/pull/20179) is ready. However, @fabioh8010 discovered that TS is very slow when integrated on VSCode. Needs investigation here to see if this is an issue that happens to everybody.
- The review for [Onyx core data + TS PR](https://github.com/blazejkustra/App/pull/1) progresses a lot thanks to @neil-marcellini. What is the time estimate for this PR to be merged?

Once the above two PRs are merged, we can execute [this script](https://github.com/Expensify/Web-Expensify/pull/37920) to create migration GH issues and officially kick off the TS migration.

### Side notes
I realized there are additional libraries being developed right now at Expensify for App

- https://github.com/Expensify/react-native-x-maps
- https://github.com/Expensify/react-native-x-draggable-list

We probably have to rewrite them in TypeScript or write TS definitions for them at some point. Just a note. I'll keep track of the related project and update here.

**Comment by blazejkustra** [comment]
> The review for https://github.com/blazejkustra/App/pull/1 progresses a lot thanks to @neil-marcellini. What is the time estimate for this PR to be merged?

@hayata-suenaga There is a TODO list in the description of the [PR](https://github.com/blazejkustra/App/pull/1#top) explaining what’s left. Currently the only blocker is react-native-onyx [PR](https://github.com/Expensify/react-native-onyx/pull/267), once it's merged, I’ll install the typed version of Onyx and the PR should be ready.

**Comment by hayata-suenaga** [comment]
merge flow

[App TS Setup PR](https://github.com/Expensify/App/pull/20179) ->

[Onyx + TS PR](https://github.com/Expensify/react-native-onyx/pull/267) -> [Core Onxy Data in App](https://github.com/blazejkustra/App/pull/1)

**Comment by hayata-suenaga** [comment]
###Weekly Update

- I created [a GH issue](https://github.com/Expensify/App/issues/23809) to dump notes on things we have to remember during the migration. Please feel free to post anything as notes to the issue.

- [ ] A performance issue with TypeScript Intellisense was discovered in [the App TS Setup PR](https://github.com/Expensify/App/pull/20179). One source of the issue was identified and fixed. @blazejkustra is working on the second source of the issue. The latest update is [here](https://github.com/Expensify/App/pull/20179#issuecomment-1655679472).

- [ ] Fabio has done amazing work so far with [Onyx + TS PR](https://github.com/Expensify/react-native-onyx/pull/267). He's [finalizing on some details of the PR](https://github.com/Expensify/react-native-onyx/pull/267#issuecomment-1656044198) but it's very close to the finish line 🥳 

- [ ] @Blajez addresses all PR reviews swiftly and [Onyx core data PR](https://github.com/blazejkustra/App/pull/1) is ready. Please give a final look and approve it if it looks good. The PR is, however, on hold for #20179 (the App TS Setup PR).
  
  Also, Blajez, the PR is for your own branch. could you make a PR for the main branch of App?
  <img width="400" alt="Screenshot 2023-07-28 at 10 51 28 AM" src="https://github.com/Expensify/App/assets/98560306/777520a4-273d-49c0-a344-6bf6a09fcb08"

**Comment by neil-marcellini** [comment]
Heads up, I'm going to be OOO Monday and Tuesday next week so please don't let me be a blocker for anything while I'm out.

**Comment by roryabraham** [comment]
Update from @blazejkustra today [here](https://expensify.slack.com/archives/C0593T50UHG/p1690818355943679)

**Comment by blazejkustra** [comment]
### Mid-week update
* [App TS Setup PR](https://github.com/Expensify/App/pull/20179) was merged yesterday 🚀 (there was a [regression reported](https://expensify.slack.com/archives/C049HHMV9SM/p1690954042805659), a [fix](https://github.com/Expensify/App/pull/24036) is already merged)
* [Add type declarations PR](https://github.com/Expensify/react-native-onyx/pull/267) is under review, @fabioh8010 is making the final touches. 
* Original [Onyx type core data](https://github.com/blazejkustra/App/pull/1) is now recreated on `Expensify/App`, here is a brand new [PR](https://github.com/Expensify/App/pull/24054). It is blocked by https://github.com/Expensify/react-native-onyx/pull/267. 

After `react-native-onyx` & `Onyx type core data` PRs are merged, we can proceed to phase 2 of Typescript migration! 🚀

**Comment by hayata-suenaga** [comment]
I didn't do end of week update last week 😭 but it's already time for...

### Mid-week update

- [Onyx PR: Create definition files](https://github.com/Expensify/react-native-onyx/pull/267)
  Thank you @fabioh8010 to really dig into TypeScript and type Onyx as strictly as possible.
  I think this PR is ready to be merged. Just checking we can actually merge this in slack.
- [App PR: Type core data](https://github.com/Expensify/App/pull/24054)
  This PR has been reviewed. Thank you @blazejkustra for handling this big PR 🙇 
  As @blazejkustra described [here](https://github.com/Expensify/App/pull/24054) and as I described in [this slack message](https://expensify.slack.com/archives/C0593T50UHG/p1691597342095349), there was a major update in Onyx and the version is being bumped in App in [this PR](https://github.com/Expensify/App/pull/24054). So we have to hold the App Setup PR for it (because the App Setup PR increases the Onyx version further).

**Comment by hayata-suenaga** [comment]
### End of week update

- Onyx PR: Create type declaration files
  Merged 🥳 🥳 🥳 ! Thank you so much @fabioh8010 and @blazejkustra for your amazing work on this one
  
- [App PR: Type core data](https://github.com/Expensify/App/pull/24054)
  As @blazejkustra described https://github.com/Expensify/App/pull/24054 and as I described in [this slack message](https://expensify.slack.com/archives/C0593T50UHG/p1691597342095349), there was a major update in Onyx and the version is being bumped in App in https://github.com/Expensify/App/pull/24054 which changes APP code to work with the major change in Onyx. So we have to hold our PR for it.

Once the App Setup PR is merged, then we're officially and finally ready to start migration.

Next week, we gonna run [this PHP script](https://github.com/Expensify/Web-Expensify/pull/37920) (internal link) with the updated CSV file that @fabioh8010 is going to provide to create migration issues.

In the initial phase of the migration:
- [SWM is planning to assign one more engineer](https://expensify.slack.com/archives/C0593T50UHG/p1691658549435879?thread_ts=1691598048.743759&cid=C0593T50UHG)
- [CallStack is planning to assign two more engineers](https://expensify.slack.com/archives/C0593T50UHG/p1691665332181349?thread_ts=1691598048.743759&cid=C0593T50UHG)
  
[SWM and CallStack are going to coordinate with each other to assign issues to themselves.](https://expensify.slack.com/archives/C0593T50UHG/p1691657781879339?thread_ts=1691598048.743759&cid=C0593T50UHG)

We have to make sure that engineers from these organizations have permissions to assign themselves to issues on GH.

**Comment by hayata-suenaga** [comment]
### Quick Update

[Here](https://github.com/Expensify/App/pull/24041#issuecomment-1679590044) is the latest update in [the issue](https://github.com/Expensify/App/pull/24041) that is blocking @blazejkustra's [Onyx core data App PR](https://github.com/Expensify/App/pull/24054). The blocking PR hasn't been merged yet.

Once @fabioh8010 returns from OOO, we can run [this PHP script](https://github.com/Expensify/Web-Expensify/pull/37920) (internal link) to create GH issues for migration tasks with the updated CSV file that @fabioh8010 is going to provide us.

**Comment by hayata-suenaga** [comment]
### End of week update

The PR that has been blocking @blazejkustra's [Onxy core data type App PR](https://github.com/Expensify/App/pull/24054) has finally been merged to main. @blazejkustra, please update the onyx import in `package.json` to use the version number. This is [the PR](https://github.com/Expensify/react-native-onyx/pull/299) that bumped the onyx version after @fabioh8010's [Onyx type definition PR](https://github.com/Expensify/react-native-onyx/pull/267) was merged. The version number we should use is `1.0.60`. The `package.json` on main branch actually specifies the version `"react-native-onyx": "1.0.61"`, so we can keep this version specification.

The migration issues were created and are tracked in [the GH TS migration project](https://github.com/orgs/Expensify/projects/46). [This is the spread sheet](https://docs.google.com/spreadsheets/d/1_ecwCjOckGzeCZYnQ4vsG7FuLV-vQIy1KG5XXWWhXpE/edit#gid=1355801134) that lists out all issues (thanks @fabioh8010 🎉). SWM and CallStack engineers, please follow the instruction specified at the top of the spread sheet to allocate issues.

Internal engineers, we have to assign engineers from these agencies to each issue as they don't have write access to GH issues. The `assigned` column in the above spread sheet keep track of issues where an engineer has already been assigned. If you see a row where `Agency Assigned` column is filled but the `assigned` column is not checked, the engineer needs to be assigned to the issue by an internal engineer. If you go to the issue, there should be a CallStack or SWM engineer who commented on the issue. Assign that engineer to the issue manually.

As soon as [this PR](https://github.com/Expensify/App/pull/24054) is merged, SWM and CallStack engineers can start working on TS migration issues 🚀

**Comment by hayata-suenaga** [comment]
Also, these internal PRs should be ready when @blazejkustra's [Onyx core data App PR](https://github.com/Expensify/App/pull/24054) is merged.

- https://github.com/Expensify/Web-Expensify/pull/38290
- https://github.com/Expensify/Auth/pull/8356

**Comment by hayata-suenaga** [comment]
## End of Week Update

We officially kicked off the migration 🎉 

Engineers from both SWM and CallStack were assigned to 47 issues so far! Several migration PRs had been already merged 💯 

<img width="500" alt="Screenshot 2023-08-25 at 7 24 22 AM" src="https://github.com/Expensify/App/assets/98560306/27d6b57b-3b6f-452b-ab05-061ee9ea293c">

### This is the official process so far that we're following for TypeScript PRs

1. A SWM or CallStack engineer picks up a migration issue from [the Google sheet](https://docs.google.com/spreadsheets/d/1GVp81ixTZ3TIspkTsgfhSA72KZGHDJlwRa4J8TtCQU4/edit?usp=sharing).
2. They put their agency and personal names on the corresponding row on the sheet. They also comment on the GH issue.
3. An internal engineer assigns them to the migration issue (SWM and CallStack engineers do NOT have to wait to be officially assigned to the issue to open a PR.
4. They open a PR and assign the Puller Bear for review. They also make sure that an engineer from the other agency is assigned to the PR as a reviewer.
5. Puller Bear auto-assigns a Contributor Plus (C+) member. C+ member does NOT review the PR but just approves the PR so that an internal engineer can be auto-assigned. C+ won't be paid. (THIS IS RELEVANT)
6. After an engineer from the other agency and an internal engineer approve the PR, the internal engineer merges the PR.
7. The migration issue is closed when the PR is deployed to production. There is no payment.
The entire process does not involve C+ (except just approving PR) or Bug Zero team members.

### For internal engineers

I think we have soft merge freeze right now, but we decided to not to apply this to TS migration PRs. TS migration RPs can be merged normally.

**Comment by hayata-suenaga** [comment]
@MelvinBot went crazy

**Comment by hayata-suenaga** [comment]
I gonna change the priority of this issue to weekly and change the frequency of updates to weekly aslo

**Comment by roryabraham** [comment]
Migration is underway! I just went through the [project board](https://github.com/orgs/Expensify/projects/46/views/1) and categorized all issues correctly, closing any issues where the migration is already done. The only thing I did not do was go through all the `TODO` issues and figure out exactly which ones need to be on `HOLD` and which are ready to be worked on.

Here's the progress so far:

|_status_|_count_|
|---|---|
|HOLD (for sure)|5|
|TODO (or maybe HOLD)|583|
|In Progress|18|
|Done|99|

It's true we have a long way to go, but we're ~20% of the way through. Let's keep pushing this forward!

**Comment by roryabraham** [comment]
Progress update:

|_status_|_count_|
|---|---|
|HOLD (for sure)|5|
|TODO (or maybe HOLD)|569|
|In Progress|21|
|Done|110|

**Comment by neil-marcellini** [comment]
Seems like the migration is cruising along

**Comment by blazejkustra** [comment]
Posted a weekly update on the slack channel [here](https://swmansion.slack.com/archives/C0593T50UHG/p1696601562024459) 🚀

**Comment by neil-marcellini** [comment]
Not overdue

**Comment by neil-marcellini** [comment]
@blazejkustra will you post another weekly update today or tomorrow?

**Comment by blazejkustra** [comment]
@neil-marcellini Yes, I'll post it on every Friday.

**Comment by blazejkustra** [comment]
New [weekly update](https://swmansion.slack.com/archives/C0593T50UHG/p1697820419846789) on the slack channel 🎉

**Comment by blazejkustra** [comment]
@fabioh8010 posted a [weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1698422636984309) on the slack channel!

**Comment by neil-marcellini** [comment]
Remove overdue label

**Comment by blazejkustra** [comment]
[weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1699025457906149) on the slack channel 🎉

**Comment by roryabraham** [comment]
Weekly progress check-in:

|Hold|TODO|In progress|Done|
|---|---|---|---|
|(at least) 4|464|39|198|

i.e: 28% complete

**Comment by roryabraham** [comment]
Implementation is chugging along. Thanks @fabioh8010 for the update!

**Comment by blazejkustra** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1700843125809069) on the slack channel ❤️ 

```
Total JS / TS files: 1838 -> 1832
JS files: 1127 -> 1094
TS files: 711 -> 738
Percentage of files migrated: 38% -> 40%
```

**Comment by neil-marcellini** [comment]
Not overdue, see update ^

**Comment by blazejkustra** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1701443751651839) on the slack channel ❤️
```
Total JS / TS files: 1832 -> 1856
JS files: 1094 -> 1045
TS files: 738 -> 811
Percentage of files migrated: 40% -> 43%
```

**Comment by neil-marcellini** [comment]
Not overdue, update was 4 days ago.

**Comment by hayata-suenaga** [comment]
last weekly update [here](https://expensify.slack.com/archives/C0593T50UHG/p1702053008749369)

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1702664213567929) on the Slack channel 🚀 

**Total JS / TS files:** 1915 -> 1945
**JS files:** 1005 -> 966
**TS files:** 910 -> 979
**Percentage of files migrated:** 47% -> 50%

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1703258212002469) on the Slack channel 🎄❤️ 

**Total JS / TS files:** 1945 -> 1967
**JS files:** 966 -> 939
**TS files:** 979 -> 1028
**Percentage of files migrated:** 50% -> 52%

**Comment by neil-marcellini** [comment]
See the last update Melvin ^

**Comment by neil-marcellini** [comment]
❗ 🔥 Heads up, I'm going to be OOO working 0% from 1/1-1/12/24. I'll be back on 1/15/24. I'm planning to work 50% tomorrow 12/29. Please feel free to un-assign, re-assign, whatever. When I return I'll pick back up whatever I can and do my best to GSD in order of priority. I will also post all of my assigned issues in #engineering-chat in Slack to try to recruit volunteers.

**Comment by blazejkustra** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1703882384427509) on the Slack channel ✨ 

```
Total JS / TS files: 1967 -> 1980
JS files: 939 -> 922
TS files: 1028 -> 1058
Percentage of files migrated: 52% -> 53%
```

**Comment by blazejkustra** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1704483690068629) on the Slack channel ✨
```
Total JS / TS files: 1980 -> 1996
JS files: 922 -> 878
TS files: 1058 -> 1118
Percentage of files migrated: 53% -> 56%
```

**Comment by hayata-suenaga** [comment]
the last update is [here](https://github.com/Expensify/App/issues/16665#issuecomment-1879175382)

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1705079918845969) on the Slack channel 🔥 

**Total JS / TS files:** 1996 -> 2011
**JS files:** 878 -> 830
**TS files:** 1118 -> 1181
**Percentage of files migrated:** 56% -> 58%

**Comment by hayata-suenaga** [comment]
[this is the last weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1705079918845969)

**Comment by roryabraham** [comment]
[latest weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1705688413297439), we hit 60% TS

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1706295868064789) on the Slack channel ❤️ 

**Total JS / TS files:** 2019 -> 2157
**JS files:** 796 -> 738
**TS files:** 1223 -> 1419
**Percentage of files migrated:** 60% -> 65%

**Comment by blazejkustra** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1706894026443639) on the Slack channel! 🐐

**Comment by hayata-suenaga** [comment]
action times are listed [here](https://expensify.slack.com/archives/C0593T50UHG/p1706894027836009)

**Comment by neil-marcellini** [comment]
We're 70% done! See that latest update from Friday [here](https://expensify.slack.com/archives/C0593T50UHG/p1707503139027929)

**Comment by blazejkustra** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1708100767320339) on the Slack channel! 💯

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1708708922572579) on the Slack channel 🔥 

**Total JS / TS files:** 2386 -> 2409
**JS files:** 597 -> 539
**TS files:** 1789 -> 1870
**Percentage of files migrated:** 74% -> 77%
**Percentage of files migrated in Onyx:** 9% -> 26%

**Comment by hayata-suenaga** [comment]
77% done 🎉

**Comment by blazejkustra** [comment]
It's Friday already, weekly update time!

**Total JS / TS files:** 2409 -> 2451
**JS files**: 539 -> 488
**TS files**: 1870 -> 1963
**Percentage of files migrated**: 77% -> 80%
**Percentage of files migrated in Ony**x: 26%

**Comment by neil-marcellini** [comment]
Not overdue, see the last update

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1709917492004849) on the Slack channel 🤩 

**Total JS / TS files:** 2451 -> 2453
**JS files:** 488 -> 429
**TS files:** 1963 -> 2024
**Percentage of files migrated:** 80% -> 82%
**Percentage of files migrated in Onyx:** 26% -> 41%

**Comment by hayata-suenaga** [comment]
weekly update [above](https://github.com/Expensify/App/issues/16665#issuecomment-1986081776)

**Comment by neil-marcellini** [comment]
❗❗ Heads up, I'm going to be OOO working from Spain 🇪🇸 part time until 3/28. Most days I will be working 50%, some days 100%. Please DM me if something needs urgent attention.❗❗

**Comment by blazejkustra** [comment]
Weekly update! We estimate that we fill finish this project until end of March ❤️ 

Total JS / TS files: 2453 -> 2516
JS files: 429 -> 349
TS files: 2024 -> 2167
Percentage of files migrated: 82% -> 86%
Percentage of files migrated in Onyx: 41% -> 53%

**Comment by roryabraham** [comment]
Still on track to finish this project by end of March? That's exciting, only 2 weeks left!

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1711138836591559) on the Slack channel 🔥 

**Total JS / TS files:** 2516 -> 2481
**JS files:** 349 -> 197
**TS files:** 2167 -> 2284
**Percentage of files migrated:** 86% -> 92% :tada:
**Percentage of files migrated in Onyx:** 53% -> 54%

**Comment by fabioh8010** [comment]
> Still on track to finish this project by end of March? That's exciting, only 2 weeks left!

I think it will extend one or two weeks due to some new TS issues that will need working, and current ones that are on hold for something, but we will probably be very close to 100% until the month ends.

**Comment by blazejkustra** [comment]
We are very close 🤌 update [here](https://swmansion.slack.com/archives/C0593T50UHG/p1711724563214139)

Total JS / TS files: 2481 -> 2523
JS files: 197 -> 100
TS files: 2284 -> 2423
Percentage of files migrated: 92% -> 96% 🚀 
Percentage of files migrated in Onyx: 54% -> 59%

**Comment by neil-marcellini** [comment]
Exciting! Not overdue.

**Comment by fabioh8010** [comment]
[Weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1712338118718959) on the Slack channel 🏎️ 🏁

**Total JS / TS files:** 2523 -> 2556
**JS files:** 100 -> 56
**TS files:** 2423 -> 2500 :tada:
**Percentage of files migrated:** 96% -> 97%
**Percentage of files migrated in Onyx:** 59% -> 57%

**Comment by roryabraham** [comment]
`useOnyx` example usages were merged yesterday, which should significantly simplify our TS usage in many of our React components. There was also some [internal discussion](https://expensify.slack.com/archives/C03TQ48KC/p1712847249274979) earlier today about some of our TS patterns causing a bit of confusion internally. Hopefully `useOnyx` alleviates some of that.

Other than that, we are very nearly done with the TS migration 🚀 it looks like the most active development is happening in Onyx [here](https://github.com/Expensify/react-native-onyx/pull/523)

**Comment by blazejkustra** [comment]
Last weekly update is on the Slack channel! We hit 100% Typescript today 🚀 

**Total JS / TS files: 2523 -> 2586
JS files: 100 -> 0
TS files: 2423 -> 2586
Percentage of files migrated: 96% -> 100%
Percentage of files migrated in Onyx: 57% -> 57%**

**Comment by hayata-suenaga** [comment]
Changing the priority of the following cleanup issues to weekly as the TS migration in the App has finished 🚀 🎉 

https://github.com/Expensify/App/issues/23499
https://github.com/Expensify/App/issues/20592
https://github.com/Expensify/Expensify/issues/290295

**Comment by fabioh8010** [comment]
[Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1713545649360149) on the Slack channel!

- **Percentage of files migrated in E/App:** 100%
- **Percentage of files migrated in Onyx:** 57% -> 61%

**Comment by neil-marcellini** [comment]
Still migrating Onyx, expecting the next update on Friday

**Comment by neil-marcellini** [comment]
The external project leaders were OOO last week so we didn't get the typical weekly update, but I expect that to resume this week.

**Comment by fabioh8010** [comment]
[Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1715355653745319) on the Slack channel!

**Comment by neil-marcellini** [comment]
^

**Comment by blazejkustra** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1715960122097649) on the Slack channel! 🚀

**Comment by roryabraham** [comment]
We are approaching the point where we should do a project wrap-up here. E/App is done, Onyx is done, only two more PRs to go in expensify-common.

**Comment by fabioh8010** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1717781330299629) on the Slack channel! 🚀

**Comment by hayata-suenaga** [comment]
I'll be out of the office (OOO) from June 10th to June 13th and will return on Friday. I think I'm not needed anymore, but I just wanted to inform you of my upcoming absence. 😄

**Comment by blazejkustra** [comment]
[Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1718379741002489) on the Slack channel!

**Comment by neil-marcellini** [comment]
Thanks!

**Comment by fabioh8010** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1718985749362139) on the Slack channel! 🚀

**Comment by roryabraham** [comment]
only a few issues remain on the project board, and I've drafted the project wrap up

**Comment by fabioh8010** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1719595933369279) on the Slack channel! 🚀

**Comment by hayata-suenaga** [comment]
the weekly update link is ⬆️

**Comment by roryabraham** [comment]
I went ahead and emailed out the [project wrap-up](https://docs.google.com/document/d/1LNcgYeYyovQ88gNjLIgBqBReLtbLAbxeAbna9QKxeAQ/edit#bookmark=id.oqwcgorv2u1i). The last things couple items are still lingering, but 99.9% of this project is done

**Comment by roryabraham** [comment]
One of the last two wrap-up issues was rejected by maintainer so we closed the issue as not planned. That leaves only https://github.com/Expensify/App/issues/43055

**Comment by fabioh8010** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1722619493325959) on the Slack channel!

**Comment by neil-marcellini** [comment]
We're incredibly close to finishing this for good, not overdue!

**Comment by neil-marcellini** [comment]
Head's up, I'll be working 50% for the rest of this week and next 🌲

**Comment by roryabraham** [comment]
@ShridharGoel reports that the last PR to close this out will be ready for review in the next couple days

**Comment by fabioh8010** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1723833511605789) on the Slack channel!

**Comment by roryabraham** [comment]
Just merged https://github.com/Expensify/eslint-config-expensify/pull/114 earlier today

**Comment by roryabraham** [comment]
Apparently [some delays for personal reasons](https://github.com/Expensify/App/pull/47439#issuecomment-2320578733) on the part of contributor for the last PR.

hopefully we can close this out by EOW 🤞🏼

**Comment by neil-marcellini** [comment]
Still waiting on that last one.

**Comment by fabioh8010** [comment]
Another [Post-migration weekly update](https://expensify.slack.com/archives/C0593T50UHG/p1726247566017509) on the Slack channel!

**Comment by neil-marcellini** [comment]
Here's the [latest update](https://expensify.slack.com/archives/C0593T50UHG/p1726843325167019).

@roryabraham it seems like it's going to take quite a while before withOnyx is deprecated. Can we sent the project wrap up sooner?

**Comment by roryabraham** [comment]
I already sent the project-wrap up for this project. The only thing left in this project is https://github.com/Expensify/App/issues/43055

**Comment by roryabraham** [comment]
I can track the official close of the project in https://github.com/Expensify/Expensify/issues/413336. We can close this issue. Thanks so much everyone for all your hard work and consistent communication. This was a huge project, and it's already hard to remember what it was like before we used TypeScript.

Huge thanks in particular to @fabioh8010 and @blazejkustra, who were the real champions of this project.
